test/src/main/java/org/axonframework/test/fixture/AxonTestFixture.java

```diff
@@ -0,0 +1,594 @@
+            this.fieldFilters.add(fieldFilter);
+            return this;
+        }
+
+        public Customization registerIgnoredField(Class<?> declaringClass, String fieldName) {
+            return registerFieldFilter(new IgnoreField(declaringClass, fieldName));
+        }
+    }
+
+    /**
+     * Every given method invocation spawn new unit of work.
+     */
+    static class Given implements AxonTestPhase.Given {
+
+        private final NewConfiguration configuration;
+        private final Customization customization;
+        private final RecordingCommandBus commandBus;
+        private final RecordingEventSink eventSink;
+        private final MessageTypeResolver messageTypeResolver;
+        private final List<AsyncUnitOfWork> unitsOfWork = new ArrayList<>();
+
+        Given(
+                NewConfiguration configuration,
+                Customization customization,
+                RecordingCommandBus commandBus,
+                RecordingEventSink eventSink,
+                MessageTypeResolver messageTypeResolver
+        ) {
+            this.configuration = configuration;
+            this.customization = customization;
+            this.commandBus = commandBus;
+            this.eventSink = eventSink;
+            this.messageTypeResolver = messageTypeResolver;
+        }
+
+        @Override
+        public AxonTestPhase.Given noPriorActivity() {
+            return this;
+        }
+
+        @Override
+        public AxonTestPhase.Given event(Object payload, MetaData metaData) {
+            var eventMessage = toGenericEventMessage(payload, metaData);
+            return events(eventMessage);
+        }
+
+        private GenericEventMessage<Object> toGenericEventMessage(Object payload, MetaData metaData) {
+            var messageType = messageTypeResolver.resolve(payload);
+            return new GenericEventMessage<>(
+                    messageType,
+                    payload,
+                    metaData
+            );
+        }
+
+        @Override
+        public AxonTestPhase.Given events(List<?>... events) {
+            var messages = Arrays.stream(events)
+                                 .map(e -> e instanceof EventMessage<?> message
+                                         ? message
+                                         : toGenericEventMessage(e, MetaData.emptyInstance())
+                                 ).toArray(EventMessage<?>[]::new);
+            return events(messages);
+        }
+
+        @Override
+        public AxonTestPhase.Given events(EventMessage<?>... messages) {
+            inUnitOfWorkRunOnInvocation(processingContext -> eventSink.publish(processingContext,
+                                                                               messages));
+            return this;
+        }
+
+        private AsyncUnitOfWork inUnitOfWorkRunOnInvocation(Consumer<ProcessingContext> action) {
+            var unitOfWork = new AsyncUnitOfWork();
```

Comment:
Why not execute them immediately? This will make sure the exception is propagated to the right line of code

Reply:
:thinking:  curious, do we make everything immediate in the test fixtures, or do we align with how the user would draft their handlers perhaps.

Reply:
I think we may execute them immediately.
Could you elaborate more on `do we align with how the user would draft their handlers perhaps` this part? I think I dont get it.

Reply:
Let's a assume the user makes a command handler like so:
```java
@CommandHandler
public CompletableFuture<Void> handle(SubscribeStudentCommand command) {
    ...
}
```

Or one like this:
```java
@CommandHandler
public Mono<Void> handle(SubscribeStudentCommand command) {
    ...
}
```

We support this with the `MessageStream` API as we know.
However, if somebody consciously makes an asynchronous method, wouldn't they expect the exception to come during subscription instead of right away?

Reply:
We had a short in-call discussion. 
When using the "given" call, you expect the message to have been handled at a point in the past before the message currently being tested.
The async nature of the handler will still lead to a unit of work, that needs to be completed at some point.
For now it seems better to us to execute the UoW immediately, so the error is thrown on the correct line.